README.md
# cielo_silent_order_post

A Flutter package that helps to use Cielo's Silent Order Post.

## Usage

These operations must be performed using its specific key (Merchand Id) in the respective environment endpoints.

```dart
var sop = SilentOrderPost(merchantId: "YOUR-MERCHANT-ID", environment: Environment.SANDBOX);
var result = await sop.sendCardData(
  SilentOrderPostRequest(
    holderName: "Maurici Ferreira Junior",
    rawNumber: "4111111111111111",
    expirationDate: "01/2030",
    securityCode: "123",
    enableBinQuery: false,
  )
);

print("STATUS-CODE: ${result?.statusCode}");

if (result?.response != null) {
  print("PAYMENT-TOKEN: ${result.response.paymentToken}");
}

if (result?.error != null) {
  if (result.error.message != null) {
    print("ERROR-MESSAGE: ${result.error.message}");
  }
  
  if (result.error.modelState != null) {
    print("VALIDATIONS:");
    if (result.error.modelState.holderName.isNotEmpty) {
      print("  - HOLDER-NAME:");
      result.error.modelState.holderName.forEach((message) {
        print("    - $message");
      });
    }
    if (result.error.modelState.rawNumber.isNotEmpty) {
      print("  - RAW-NUMBER:");
      result.error.modelState.rawNumber.forEach((message) {
        print("    - $message");
      });
    }
    if (result.error.modelState.expiration.isNotEmpty) {
      print("  - EXPIRATION-DATE:");
      result.error.modelState.expiration.forEach((message) {
        print("    - $message");
      });
    }
    if (result.error.modelState.securityCode.isNotEmpty) {
      print("  - SECURITY-CODE:");
      result.error.modelState.securityCode.forEach((message) {
        print("    - $message");
      });
    }
  }
}
```


CHANGELOG.md
## 1.0.2

* Bug fixes and code improvements

## 1.0.1

* Fixed map on null on error response

## 1.0.0

* Initial version of the library

test/cielo_silent_order_post_test.dart
import 'package:cielo_silent_order_post/src/silent_order_post_request.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:cielo_silent_order_post/cielo_silent_order_post.dart';

void main() {
  String correctMerchantId = "YOUR-MERCHANT-ID";
  String incorrectMerchantId = "ZZZZZZZZ-ZZZZ-ZZZZ-ZZZZ-ZZZZZZZZZZZZ";
  SilentOrderPostRequest correctRequest = SilentOrderPostRequest(
    holderName: "Maurici Ferreira Junior",
    rawNumber: "41111111111111111",
    expirationDate: "01/2030",
    securityCode: "123",
    enableBinQuery: false,
  );
  SilentOrderPostRequest incorrectRequest = SilentOrderPostRequest(
    holderName: "Maurici F Junior 1234567890",
    rawNumber: "AAAAAAAAAAAAAAAAA",
    expirationDate: "BB/CCCC",
    securityCode: "DDD",
    enableBinQuery: false,
  );

  group("Correct merchant id and correct request", () {
    SilentOrderPost sop;
    SilentOrderPostResult result;

    setUp(() async {
      sop = SilentOrderPost(merchantId: correctMerchantId, environment: Environment.SANDBOX);
      result = await sop.sendCardData(correctRequest);
    });

    test("should return status code 201 created", () {
      expect(result?.statusCode, 201);
    });

    test("should return success response", () {
      expect(result, isNotNull);
      expect(result?.response, isNotNull);
      expect(result?.response?.paymentToken, isNotNull);
    });

    test("should return null error response", () {
      expect(result?.error, isNull);
    });
  });

  group("Incorrect merchant id", () {
    SilentOrderPost sop;
    SilentOrderPostResult result;

    setUp(() async {
      sop = SilentOrderPost(merchantId: incorrectMerchantId, environment: Environment.SANDBOX);
      result = await sop.sendCardData(correctRequest);
    });

    test("should return status code 401 unauthorized", () {
      expect(result?.statusCode, 401);
    });

    test("should return error response", () {
      expect(result, isNotNull);
      expect(result?.error, isNotNull);
      expect(result?.error?.message, "unauthorized");
    });
  });

  group("Correct merchant id and incorrect card data", () {
    SilentOrderPost sop;
    SilentOrderPostResult result;

    setUp(() async {
      sop = SilentOrderPost(merchantId: correctMerchantId, environment: Environment.SANDBOX);
      result = await sop.sendCardData(incorrectRequest);
    });

    test("should return error response", () {
      expect(result, isNotNull);
      expect(result?.error, isNotNull);
      expect(result?.error?.modelState, isNotNull);
    });

    test("should return model state with validations", () {
      expect(result?.error?.modelState?.holderName, isNotEmpty);
      expect(result?.error?.modelState?.rawNumber, isNotEmpty);
      expect(result?.error?.modelState?.expiration, isNotEmpty);
      expect(result?.error?.modelState?.securityCode, isNotEmpty);
    });
  });

  group("Bin Query flag set as true", () {
    SilentOrderPost sop;
    SilentOrderPostResult result;
    correctRequest.enableBinQuery = true;

    setUp(() async {
      sop = SilentOrderPost(merchantId: correctMerchantId, environment: Environment.SANDBOX);
      result = await sop.sendCardData(correctRequest);
    });

    test("should return bin query result", () {
      expect(result, isNotNull);
      expect(result?.response, isNotNull);
      expect(result?.response?.binQueryReturnMessage, isNotNull);
      expect(result?.response?.binQueryReturnCode, isNotNull);
      expect(result?.response?.cardBin, isNotNull);
      expect(result?.response?.cardLast4Digits, isNotNull);
      expect(result?.response?.foreignCard, isNotNull);
    });
  });
}


lib/cielo_silent_order_post.dart
library cielo_silent_order_post;

import 'dart:convert';

import 'package:http/http.dart' as http;

import 'src/access_token_response.dart';
import 'src/environment.dart';
import 'src/silent_order_post_error_response.dart';
import 'src/silent_order_post_request.dart';
import 'src/silent_order_post_response.dart';
import 'src/silent_order_post_result.dart';

export 'src/environment.dart';
export 'src/silent_order_post_error_response.dart';
export 'src/silent_order_post_request.dart';
export 'src/silent_order_post_response.dart';
export 'src/silent_order_post_result.dart';

class SilentOrderPost {
  final String merchantId;
  final Environment environment;
  bool binQuery = false;
  bool verifyCard = false;
  bool tokenize = false;
  String _baseUrl;

  SilentOrderPost({this.merchantId, this.environment}) {
    this._baseUrl = getBaseUrl(environment);
  }

  Future<AccessTokenResult> _getAccessToken() async {
    AccessTokenResponse accessTokenResponse;

    final Uri uri = Uri.https(this._baseUrl, "/post/api/public/v1/accesstoken",
        {"merchantid": this.merchantId});
    http.Response response = await http.post(uri);

    return checkResponseAccessToken(response, accessTokenResponse);
  }

  Future<SilentOrderPostResult> sendCardData(
      SilentOrderPostRequest request) async {
    AccessTokenResult resultFromAccessToken = await _getAccessToken();

    if (resultFromAccessToken.response?.accessToken != null) {
      http.Response response =
          await sendRequest(request, resultFromAccessToken, this._baseUrl);

      return checkSendCardDataResponse(response);
    } else {
      return SilentOrderPostResult(
          error: SilentOrderPostErrorResponse(
              message: resultFromAccessToken.error),
          statusCode: resultFromAccessToken.statusCode);
    }
  }

  SilentOrderPostResult checkSendCardDataResponse(http.Response response) {
    switch (response.statusCode) {
      case 201:
        return SilentOrderPostResult(
            response:
                SilentOrderPostResponse.fromJson(jsonDecode(response.body)),
            statusCode: response.statusCode);
      case 400:
        return SilentOrderPostResult(
            error: SilentOrderPostErrorResponse.fromJson(
                jsonDecode(response.body)),
            statusCode: response.statusCode);
      case 401:
        return SilentOrderPostResult(
            error: SilentOrderPostErrorResponse(message: "unauthorized"),
            statusCode: response.statusCode);
      case 404:
        return SilentOrderPostResult(
            error: SilentOrderPostErrorResponse(message: "not_found"),
            statusCode: response.statusCode);
      default:
        return SilentOrderPostResult(
            error: SilentOrderPostErrorResponse(message: "unknown_error"),
            statusCode: response.statusCode);
    }
  }
}

String getBaseUrl(Environment environment) {
  return environment == Environment.PRODUCTION
      ? "transaction.cieloecommerce.cielo.com.br"
      : "transactionsandbox.pagador.com.br";
}

AccessTokenResult checkResponseAccessToken(
    http.Response response, AccessTokenResponse accessTokenResponse) {
  switch (response.statusCode) {
    case 201:
      accessTokenResponse =
          AccessTokenResponse.fromJson(jsonDecode(response.body));
      return AccessTokenResult(
          response: accessTokenResponse, statusCode: response.statusCode);
      break;
    case 401:
      return AccessTokenResult(
          error: "unauthorized", statusCode: response.statusCode);
    case 404:
      return AccessTokenResult(
          error: "not_found", statusCode: response.statusCode);
    default:
      return AccessTokenResult(
          error: "unknown_error", statusCode: response.statusCode);
  }
}

Future<http.Response> sendRequest(SilentOrderPostRequest request,
    AccessTokenResult resultFromAccessToken, String baseUrl) async {
  var sdkName = "CieloSilentOrderPost";
  var sdkVersion = "1.0.2";

  request.accessToken = resultFromAccessToken.response.accessToken;
  final Uri uri = Uri.https(baseUrl, "/post/api/public/v1/card");
  return await http.post(
    uri,
    headers: <String, String>{
      "Content-Type": "application/json",
      "x-sdk-version": "$sdkName-Dart@$sdkVersion"
    },
    body: jsonEncode(request.toJson()),
  );
}

class AccessTokenResult {
  AccessTokenResponse response;
  String error;
  int statusCode;

  AccessTokenResult({this.response, this.error, this.statusCode});
}


lib/src/silent_order_post_error_response.dart
class SilentOrderPostErrorResponse {
  SilentOrderPostErrorResponse({
    this.message,
    this.modelState,
  });

  String message;
  ModelState modelState;

  factory SilentOrderPostErrorResponse.fromJson(Map<String, dynamic> json) => SilentOrderPostErrorResponse(
    message: json["Message"],
    modelState: ModelState.fromJson(json["ModelState"]),
  );
}

class ModelState {
  ModelState({
    this.holderName,
    this.rawNumber,
    this.expiration,
    this.securityCode,
  });

  List<String> holderName;
  List<String> rawNumber;
  List<String> expiration;
  List<String> securityCode;

  factory ModelState.fromJson(Map<String, dynamic> json) => ModelState(
    holderName: (json["request.HolderName"] != null) ? List<String>.from(json["request.HolderName"]?.map((x) => x)) : List<String>(),
    rawNumber: (json["request.RawNumber"] != null) ? List<String>.from(json["request.RawNumber"]?.map((x) => x)) : List<String>(),
    expiration: (json["request.Expiration"] != null) ? List<String>.from(json["request.Expiration"]?.map((x) => x)) : List<String>(),
    securityCode: (json["request.SecurityCode"] != null) ? List<String>.from(json["request.SecurityCode"]?.map((x) => x)) : List<String>(),
  );
}

example/main.dart
import 'package:cielo_silent_order_post/cielo_silent_order_post.dart';

void main(List<String> arguments) async {
  var sop = SilentOrderPost(merchantId: "YOUR-MERCHANT-ID", environment: Environment.SANDBOX);
  var result = await sop.sendCardData(
      SilentOrderPostRequest(
        holderName: "Maurici Ferreira Junior",
        rawNumber: "4111111111111111",
        expirationDate: "01/2030",
        securityCode: "123",
        enableBinQuery: false,
      )
  );

  print("STATUS-CODE: ${result?.statusCode}");

  if (result?.response != null) {
    print("PAYMENT-TOKEN: ${result.response.paymentToken}");
  }

  if (result?.error != null) {
    if (result.error.message != null) {
      print("ERROR-MESSAGE: ${result.error.message}");
    }

    if (result.error.modelState != null) {
      print("VALIDATIONS:");
      if (result.error.modelState.holderName.isNotEmpty) {
        print("  - HOLDER-NAME:");
        result.error.modelState.holderName.forEach((message) {
          print("    - $message");
        });
      }
      if (result.error.modelState.rawNumber.isNotEmpty) {
        print("  - RAW-NUMBER:");
        result.error.modelState.rawNumber.forEach((message) {
          print("    - $message");
        });
      }
      if (result.error.modelState.expiration.isNotEmpty) {
        print("  - EXPIRATION-DATE:");
        result.error.modelState.expiration.forEach((message) {
          print("    - $message");
        });
      }
      if (result.error.modelState.securityCode.isNotEmpty) {
        print("  - SECURITY-CODE:");
        result.error.modelState.securityCode.forEach((message) {
          print("    - $message");
        });
      }
    }
  }
}